### PR TITLE
Fix C++ data parallel

### LINKED ISSUE #19540: C++ Frontend data_parallel Does Not Update Weights
## 🐛 Bug

I have adapted the mnist.cpp example to work with two GPUs using torch::nn::parallel::data_parallel. However, the loss does not decay, and the accuracy never improves beyond 0.114 (the original single GPU example reaches around 0.99 accuracy).

## To Reproduce

1. Compile and run the attached example (change mnist_parallel.cpp.txt to mnist_parallel.cpp).
[mnist_parallel.cpp.txt](https://github.com/pytorch/pytorch/files/3100695/mnist_parallel.cpp.txt)
[CMakeLists.txt](https://github.com/pytorch/pytorch/files/3100696/CMakeLists.txt)

## Expected behavior

I would expect that the network could improve accuracy during training, but it does not.
## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
PyTorch version: 1.0.1
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.14.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti

Nvidia driver version: 410.104
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.0
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip3] numpy==1.15.0
[conda] blas                      1.0                         mkl  
[conda] magma-cuda10              2.4.0                         1    cpbotha
[conda] magma-cuda100             2.5.0                         1    pytorch
[conda] magma-cuda90              2.5.0                         1    pytorch
[conda] mkl                       2019.3                      199  
[conda] mkl-include               2019.3                      199  
[conda] mkl-service               1.1.2            py36he904b0f_5  
[conda] mkl_fft                   1.0.10           py36ha843d7b_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] mkldnn                    0.16.1                        0    mingfeima
[conda] pytorch                   1.0.1           cuda100py36he554f03_0  
[conda] torchvision               0.2.1                    py36_0

## Additional context

If this is not a bug, and I am setting up the code incorrectly, please let me know and I will issue a feature request instead for a good data_parallel C++ example.

### PR DESCRIPTION
**Pull request by mrshenli**
Fixes #19540

CC @nmerrill67

C++ data parallel was using Module.clone() to create module replicas on every destination device. However, clone() does not set up gradient edges to point from replicas to the original module. As a result, the gradient will not be aggregated into the original module. This commit fixes the the problem by manually setting gradient edges from every parameter X in every replica to the same parameter X in the original module.

## Failed Attempt

Initially I tried implementing what we did in `replicate.py`, which
1. create module replicas
2. use Python `Broadcast` autograd function to broadcast every parameter in the original module to all destination devices.
3. assign the broadcast result params to module replicas' `_parameters` dict. 

This works in Python because derived module member field params (e.g., `Linear.weight`) and base module `_parameters` (e.g., `Linear._parameters['weight']`) are referencing the same parameter instance. Assigning one of them will apply to both. However, in C++, even though I can modify Module's `parameters_ `values and gradient edges to point to the broadcast source,  I cannot touch the weight and bias member fields in Linear, because replicate cannot (and should not) add special-case handlers to every different module. (See `Linear` [.h](https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/modules/linear.h), [.cpp](https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/src/nn/modules/linear.cpp)) Although they initially point to the same `TensorImpl` instance, after assigning to `Module.parameters_['weight']`, it will be different from `Linear.weight`.

## Solution Options

@gchanan and I had several discussions on this issue and figured two solutions to this problem.

### Option One [implemented in this PR]

Replicate the module in two steps:
1. call `Module.clone()` to create a module replica on every destination device.
2. manually setting gradient edges from every parameter in every replica to the same parameter in the original module. 

* Pro: Does not need to change any existing module, and relatively easier to implement
* Con: It is a little hackish. 

### Options Two

Implement a `Replicatable` class (similar to `Cloneable`), and make it a friend class of `Module`. For more details see `Note [Replicating Modules]` in the code change.

* Pro: Maybe this aligns more with our existing approach implemented in `Cloneable`?
* Con: Require changes to every existing module.

I am inclined to go with option one, because `replicate` will only be used on data parallel. I feel it is too big an overkill if we have to change all existing module implementations due to a data parallel requirement.

### CODE REVIEW

File: test/cpp/api/parallel.cpp [hunk 1]
```diff
@@ -17,6 +20,13 @@
 using namespace torch::autograd;
 using namespace torch::nn;
 
+template <typename T>
+bool almost_equal(torch::Tensor left, torch::Tensor right, T tolerance = 1e-4) {
```
Comment (vishwakftw) [inline test/cpp/api/parallel.cpp]: You can consider using `allclose` for this if it helps.

**Comment by ezyang** [comment]
Just replicating here what I told @mrshenli in person: if the problem is that the two copies of Tensor in Module can get out of sync in C++ but not in Python, the correct solution is to make sure they stay in sync, one way or another.

Taking a closer look now.

**Comment by ezyang** [comment]
Some clarification about how replicate.py works: 

> This works in Python because derived module member field params (e.g., Linear.weight) and base module _parameters (e.g., Linear._parameters['weight']) are referencing the same parameter instance. Assigning one of them will apply to both.

This is because there isn't actually a `Linear.weight`. It is provided by the `__getattr__` magic method on Module:

```
    def __getattr__(self, name):
        if '_parameters' in self.__dict__:
            _parameters = self.__dict__['_parameters']
            if name in _parameters:
                return _parameters[name]
```

**Comment by ezyang** [comment]
After reminding myself about how parameter registration works in C++, I now agree that it is impossible to do assignment after clone. So I think that fixing up the cloned tensor directly is the right way to go.

**Comment by SsnL** [comment]
This is completely related to this PR. But out of curiosity I checked c++ DP code. It seems that c++ DP is quite a bit slower than the python one because `module.clone()` does `.clone()` before `.to(*device)` https://github.com/pytorch/pytorch/blob/8cde4c4d223d3eb1179f87fd6336d25c189acb98/torch/csrc/api/include/torch/nn/cloneable.h#L34-L75

Moreover, it is also **incorrect** with BNs (and other things that relies on in-place changes to buffers/parameters) because `module.clone` always **deepcopies**, i.e., the buffers and parameters are always **cloned** as replica, even if the source module lives on the correct device. This seems to me that it would be impossible to correct perform BN.

**Comment by mrshenli** [comment]
@SsnL thanks for the catch. I added #20995 to remove the unnecessary clone. For BN use cases, let me revise this PR and add a test for it.

File: torch/csrc/api/include/torch/nn/parallel/data_parallel.h [hunk 2]
```diff
@@ -24,6 +25,115 @@
 
 namespace torch {
 namespace nn {
+
+namespace {
+
+// Note [Replicating Modules]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (ezyang) [inline torch/csrc/api/include/torch/nn/parallel/data_parallel.h]: Thank you for a nice note, it is much appreciated!

File: torch/csrc/api/include/torch/nn/parallel/data_parallel.h [hunk 3]
```diff
@@ -24,6 +25,115 @@
...
+    for (auto& input: inputs) {
+      TORCH_CHECK(input.sizes() == inputs[0].sizes(),
+          "All inputs of ReduceAdd must have the same size, but got ",
+          input.sizes(), " and ", inputs[0].sizes());
+
+      TORCH_CHECK(input.dtype() == inputs[0].dtype(),
+          "All inputs of ReduceAdd must have the same dtype, but got ",
+          input.dtype(), " and ", inputs[0].dtype());
+
+      // TODO: use nccl reduce
```
Comment (ezyang) [inline torch/csrc/api/include/torch/nn/parallel/data_parallel.h]: Does this TODO mean that C++ data parallel is less efficient than Python data parallel at the moment? Seems like worth a big warning in that case.

File: torch/csrc/api/include/torch/nn/parallel/data_parallel.h [hunk 4]
```diff
@@ -40,6 +150,11 @@ std::vector<std::shared_ptr<ModuleType>> replicate(
     replicas.push_back(
         std::dynamic_pointer_cast<ModuleType>(module->clone(device)));
   }
+  // Configure gradient edges to point from replcia parameters to original
+  // module parameters. See [Replicating Modules]
+  std::vector<std::shared_ptr<Module>> untyped_replicas(
+    replicas.begin(), replicas.end());
```
Comment (ezyang) [inline torch/csrc/api/include/torch/nn/parallel/data_parallel.h]: I don't see why you have to allocate a vector here. Can't you just pass the iterators to the function?

File: torch/csrc/api/include/torch/nn/parallel/data_parallel.h (hunk 3 above)
Comment (mrshenli) [reply torch/csrc/api/include/torch/nn/parallel/data_parallel.h]: Yes, the the current version would be slower. Let me implement nccl reduce and remove the TODO here.

File: torch/csrc/api/include/torch/nn/parallel/data_parallel.h (hunk 4 above)
Comment (ezyang) [reply torch/csrc/api/include/torch/nn/parallel/data_parallel.h]: I guess this is because of the way you handle children, because you concatenate the replicas of all sibling children (not just the current child) and now you don't have iterators. But that seems like a pretty unusual to write the recursion: normally you'd just recurse on child modules.  Is there a reason it was done this way?

File: torch/csrc/api/include/torch/nn/parallel/data_parallel.h [hunk 5]
```diff
@@ -24,6 +25,115 @@
...
+//
+// Module replication is implemented in the following two steps:
+// 1) create a module replica on each destination device using Module.clone().
+// 2) manually add a gradient edge pointing from every parameter X in every
+//    module replica to the same parameter X in the original module, using
+//    ReduceAdd as the grad_fn.
+//
+// ReduceAdd can ONLY be used during the backward pass of data parallel. Forward
+// pass cannot use this function as it does not setup gradient function and
+// history at all. Do NOT try to use ReduceAdd for any other purposes.
```
Comment (ezyang) [inline torch/csrc/api/include/torch/nn/parallel/data_parallel.h]: Do we support double backwards through replicate in Python? I don't see any indication in issues that we don't...

File: torch/csrc/api/include/torch/nn/parallel/data_parallel.h (hunk 4 above)
Comment (mrshenli) [reply torch/csrc/api/include/torch/nn/parallel/data_parallel.h]: I remember I run into problems when collecting children modules on ling 126. Let me try if I can drop this `untyped_replicas `.

**Comment by ezyang** [comment]
If you want to land this to fix an urgent bug, I suppose I can be convinced to land it in its current state (I agree it's an improvement), but what I am confused about is why you had to implement ReduceAdd from scratch. Ostensibly, the Python implementation of DataParallel sets up *some* autograd backwards function, and to me, it seems like you ought to use whatever that backwards function was, and not make up a new one (this would also fix your TODO, if we are using nccl reduce in Python).

**Comment by ezyang** [comment]
Hang on, before you do that, read my overall comment please :)

Excerpts from mrshenli's message of 2019-05-29 08:57:48 -0700:
> mrshenli commented on this pull request.
> 
> > +  autograd::variable_list apply(autograd::variable_list&& inputs) override {
> +    TORCH_CHECK(!compute_requires_grad(inputs),
> +        "ReduceAdd can only be used during the backward pass of data parallel.");
> +
> +    Tensor output = torch::zeros_like(inputs[0], {destination_device_});
> +
> +    for (auto& input: inputs) {
> +      TORCH_CHECK(input.sizes() == inputs[0].sizes(),
> +          "All inputs of ReduceAdd must have the same size, but got ",
> +          input.sizes(), " and ", inputs[0].sizes());
> +
> +      TORCH_CHECK(input.dtype() == inputs[0].dtype(),
> +          "All inputs of ReduceAdd must have the same dtype, but got ",
> +          input.dtype(), " and ", inputs[0].dtype());
> +
> +      // TODO: use nccl reduce
> 
> Yes, the the current version would be slower. Let me implement nccl reduce and remove the TODO here. 
>

File: torch/csrc/api/include/torch/nn/parallel/data_parallel.h (hunk 4 above)
Comment (mrshenli) [reply torch/csrc/api/include/torch/nn/parallel/data_parallel.h]: I think I can recurse on child modules by creating a map from `*parameter` to `grad_fn`, and retrieve that `grad_fn` in every recursion.

**Comment by mrshenli** [comment]
> but what I am confused about is why you had to implement ReduceAdd from scratch. Ostensibly, the Python implementation of DataParallel sets up some autograd backwards function, and to me, it seems like you ought to use whatever that backwards function was, and not make up a new one (this would also fix your TODO, if we are using nccl reduce in Python).

Python's `DataParallel` uses `Broadcast` and `ReduceAddCoalesced` functions which are implemented in Python. What about letting me create a new PR to move these two functions to C++?

File: torch/csrc/api/include/torch/nn/parallel/data_parallel.h (hunk 5 above)
Comment (mrshenli) [reply torch/csrc/api/include/torch/nn/parallel/data_parallel.h]: Haven't tried yet. Let me add a test.

**Comment by ezyang** [comment]
Yes, that seems like the right way to do it. If you want to apply this hotfix on master, I'll let it through, but ideally this should just all live in C++.

**Comment by mrshenli** [comment]
> If you want to apply this hotfix on master, I'll let it through, but ideally this should just all live in C++.

Thanks. Let me first address smaller revision requests and create issues to track bigger ones. Will do that later today or tmr. I need to finish up sth else now.

**Comment by mrshenli** [comment]
The latest commit addressed @vishwakftw's comments on using `allclose` and @ezyang's comments on removing `untyped_replicas`. I didn't modify the recursion strategy, because after a second thought I think the current one is easier to understand and cleaner, i.e., recurse based on the original module's structure instead of creating a big map from all params/buffers to `grad_fn`s upfront and then recurse on every original-replica pair. The current solution to the problem is adding a `typename ModuleType` to `replicate_grad_edges`. The template is only useful in the very first invocation, and `ModuleType` becomes `Module` in all subsequent invocations.

TODOs are in #21144.

@ezyang is the current solution acceptable?

**Review by ezyang — APPROVED** [review]
Let's go ahead and land this for now. Looking up to the follow up where we line up the Python and C++ versions exactly.

**Comment by mrshenli** [comment]
Failed test passed in [rerun](https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-devtoolset7-rocmrpm-centos7.5-trigger-test/26777/).

**Comment by yulinhuyang** [comment]
hi mrshenli！
I use my module like this in C++ :

struct ConvModel : torch::nn::Module
{
  torch::Tensor forward(at::Tensor gallery)
  {

    auto result = torch::nn::functional::conv1d(gallery, query_weight);
    return result;
  }

and i  use torch::nn::parallel::replicate

but i  error 

clone() has not been implemented for ConvModel. Subclass torch::nn::Cloneable<ConvModel> instead of torch::nn::Module to inherit the ability to clone

if I try  in herit    torch::nn::Cloneable<ConvModel> ,it have anthor error.   can you give me some suggestion about it ?@mrshenli
